src/NuGet.Clients/NuGet.Console/OutputConsole/ChannelOutputConsole.cs

```diff
@@ -158,6 +158,8 @@ protected virtual void Dispose(bool disposing)
                     {
                         // Ignore exceptions
                     }
+
+                    _pipeLock.Dispose();
```

Comment:
Concurrent dispose might make this fail. 

Consider what happens if multiple threads try to dispose this lock at the same time. Some of the methods called by dispose depend on pipelock.

Reply:
As per [docs](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.threading.asyncsemaphore?view=visualstudiosdk-2019), `This object does not need to be disposed of, as it does not hold unmanaged resources. Disposing this object has no effect on current users of the semaphore, and they can release their hold on the semaphore without exception. An ObjectDisposedException is thrown back at anyone asking to or waiting to enter the semaphore after Dispose() is called. `

`AsyncSemaphore.Dispose()` method appears thread safe because of the `lock (this.syncObject)` statement in the implementation.

https://github.com/microsoft/vs-threading/blob/fad0c9a9fe47ae0678c81f6ffd253b7a02d21b93/src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs#L179-L209

> Some of the methods called by dispose depend on pipelock.

You are correct Nikolche. Hence, I added `_pipeLock.Dispose(); ` line of code at the end of `if (disposing)` scope so that `methods invoked in Dispose() have access to the _pipeLock instance`.

Reply:
That doesn't solve the race. 

CloseChannelAsync might call pipelock.EnterAsync which would throw.

Reply:
I am okay to revert this change and add `justification` to the suppression message. I have a quick question before I consider that option.

`_pipelock` is used in `WriteToOutputChannelAsync`, `CloseChannelAsync` and `Dispose (bool disposing)` methods.

There are zero references to the `public Dispose()` method in the product code. `_pipelock` doesn't get disposed unless some external caller invokes `Dispose()` method which internally invokes `Dispose (disposing: true) `. I am curious to understand why would an `external` caller invoke methods after the `ChannelOutputConsole` object has been disposed?

![image](https://user-images.githubusercontent.com/52756182/112680941-3fcbf080-8e2b-11eb-9f57-1173fd5fac15.png)

Reply:
> I am curious to understand why would an external caller invoke methods after the ChannelOutputConsole object has been disposed?

It's more about the correctness of the dispose implementation. Today, nothing bad can happen.
Dispose has a well documented behavior, so anyone that might be disposing the channeloutputconsole should get the expected behavior.

Reply:
`?` checks for null. A disposed object won't be null.

Reply:
I'd just avoid dispose the semaphore since there are no sideeffects from it.

Reply:
Copying from my chat with @kartheekp-ms 

Say the dispose method, https://github.com/NuGet/NuGet.Client/blob/31270163e1021637bc1421cb2d44390aca6addc5/src/NuGet.Clients/NuGet.Console/OutputConsole/ChannelOutputConsole.cs#L143-L166 is called concurrently. 
 
Because the operations in the dispose a not atomic, it's possible that both threads make it to line 149. It's also possible that one of the threads reahes 155, while the other one reaches 162 already. 
When the thread on 155 finishes, it might've thrown, but I see that's caught, so shouldn't be a problem.

So there's no bad side-effects with unexpected exceptions thrown.